Add Metricbeat AWS EC2 Module

kaiyan-sheng opened the pull request:
This PR is to add AWS EC2 metricset into metricbeat. This PR enables metricbeat to periodically fetch EC2 monitoring metrics from AWS Cloudwatch using GetMetricData API. Also for each EC2 instance, it uses DescribeInstances API to get detailed information about a specific instance. 

Checklist:
* [x] Integration tests exist
* [x] System tests exist
* [x] Automated checks that all fields are documented
* [x] Documentation
* [x] Fields follow [ECS](https://github.com/elastic/ecs) and [naming conventions](https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html)
* [x] Example `data.json` exists and an automated way to generate it exists (`go test -data`)
* [x] Dashboards exists (if applicable)
* [ ] Kibana Home Tutorial (if applicable)

There are 16 cloudwatch metrics for EC2 monitoring collected by  AWS EC2 module:
CPUUtilization -> aws.ec2.cpu.total.pct
CPUSurplusCreditBalance -> aws.ec2.cpu.surplus_credits_balance
CPUSurplusCreditsCharged -> aws.ec2.cpu.surplus_credits_charged
CPUCreditBalance -> aws.ec2.cpu.credit_balance
CPUCreditUsage -> aws.ec2.cpu.credit_usage
NetworkPacketsIn -> aws.ec2.network.in.packets
NetworkPacketsOut -> aws.ec2.network.out.packets
NetworkIn -> aws.ec2.network.in.bytes
NetworkOut -> aws.ec2.network.out.bytes
DiskReadBytes -> aws.ec2.diskio.read.bytes
DiskWriteOps -> aws.ec2.diskio.write.bytes
DiskReadOps -> aws.ec2.diskio.read.ops
DiskWriteBytes -> aws.ec2.diskio.write.ops
StatusCheckFailed -> aws.ec2.status.check_failed
StatusCheckFailed_System -> aws.ec2.status.check_failed_system
StatusCheckFailed_Instance -> aws.ec2.status.check_failed_instance

Also this PR adds metric collection for basic AWS EC2 instance information:
cloud.provider
cloud.instance.id
cloud.image.id
cloud.machine.type
cloud.region
cloud.availability_zone

kaiyan-sheng commented:
I'm thinking by following https://github.com/elastic/ecs#network, rename the cpu related metrics to cpu.*, network related metrics to network.*, disk related metrics to disk.*, and status related metrics to status.*

ruflin commented:
For the metrics best follow the naming scheme that the system metricbeat module uses.

kaiyan-sheng commented:
@ruflin Yes good point, but some of the cloudwatch metrics don't have a direct match to metrics in system metricbeat module. What I found can be a direct map are:
- cpu_utilization -> cpu.total.pct
- network_packets_in -> network.in.packets
- network_packets_out -> network.out.packets
- network_in -> network.in.bytes
- network_out -> network.out.bytes
- disk_read_bytes -> diskio.read.bytes
- disk_write_bytes -> diskio.write.bytes

What do you think?

ruflin commented:
Yes, lets only map the ones that exist, the other ones which are EC2 specific I would keep prefixed.

Some of the above values are actually matching to ECS which is even nicer: https://github.com/elastic/ecs#network For the cpu / disk ones: Keep in mind that they are prefixed by `system.*`

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,330 @@
+		return nil, err
+	}
+
+	os.Setenv("AWS_ACCESS_KEY_ID", config.AccessKeyID)
```
Is there any other way to pass these? Passing by value would be best but we can probably live with this

kaiyan-sheng replied:
By default, it goes to .aws/credentials to look for credentials first.

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,330 @@
+		svcEC2 := ec2.New(cfg)
+		instanceIDs, instancesOutputs, err := getInstancesPerRegion(svcEC2)
+		if err != nil {
+			report.Error(errors.Wrap(err, "getInstancesPerRegion failed"))
```
`report.Error` won't stop this method, do we really want to continue with the other regions? at the very least we should call `continue` to skip this one

kaiyan-sheng replied:
Yeah good catch on the `continue`. I do think we want to continue with other regions though. For example I had seen regionList includes 'us-gov' before and it will return an error here since we dont have permission to visit us-gov. But the same api call works fine with other regions.

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,330 @@
+	req := svc.DescribeRegionsRequest(input)
+	output, err := req.Send()
+	if err != nil {
+		fmt.Println("Failed DescribeRegions: ", err)
```
leftover?

exekias replied:
I see there are several of these, you should use `logp.Err` where you want to show the error in the log, or just remove the Print statement if you want to pass it

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,330 @@
+	event.ModuleFields.Put("cloud.machine.type", machineType)
+	event.ModuleFields.Put("cloud.availability_zone", *instanceOutput.Placement.AvailabilityZone)
+	event.ModuleFields.Put("cloud.image.id", *instanceOutput.ImageId)
+	event.ModuleFields.Put("cloud.region", regionName)
```
All these `cloud.*` should be Put under `event.RootFields`, so they end up matching with what `add_cloud_metadata` provides

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,330 @@
+			continue
+		}
+		switch *output.Id {
+		case "cpu1":
```
No need to change it, but as this list grows, it sounds like having a translating map like:
```
map[string]string{
  "cpu1": "cpu.total.pct",
  "cpu2": "cpu.credit_usage",
  ...
}
```

could save you some code

kaiyan-sheng replied:
Great idea!! Thanks!! I don't like that big switch case statement :-)

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,330 @@
+func getMetricDataPerRegion(instanceID string, nextToken *string, svc cloudwatchiface.CloudWatchAPI) (*cloudwatch.GetMetricDataOutput, error) {
+	//TODO:remove hard coded variables
+	endTime := time.Now()
+	duration, err := time.ParseDuration("-10m")
```
could you please add a comment explaining the `"-10m"` value?

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,330 @@
+	endTime := time.Now()
+	duration, err := time.ParseDuration("-10m")
+	if err != nil {
+		fmt.Println("Error ParseDuration: ", err)
```
leftover, you can also wrap the error

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,330 @@
+		Value: &instanceID,
+	}
+
+	metricDataQuery1 := createMetricDataQuery("cpu1", "CPUUtilization", []cloudwatch.Dimension{dim})
```
I see the map would be more complex if you include this too, but can be helpful

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,330 @@
+
+	startTime := endTime.Add(duration)
+
+	//TODO:add function getMetricNames from environment variables
```
what would this `getMetricNames` do?

kaiyan-sheng replied:
Haha I thought I saw a getMetricName method somewhere in aws api to query all the available metric names from cloudwatch ec2. But I guess it's not necessary here and probably will make the code even messier.

exekias commented:
@ruflin about metric names, as there is no direct map with `system` module, I think that the current approach (names like `aws.ec2.cpu.total.pct`) is correct. As a user that will help me find `ec2` related metrics while typing them in the bar.

At some point we will also want this to work with current Waffle map view and others, let's review the metrics that are being used there and see our options. For instance, we could put those under `system.*` namespace and create an alias to have them under `aws.ec2.*` too

ruflin commented:
@exekias I like the idea to start to play with aliases here. We will have to make sure that the units etc. are all in sync.

kaiyan-sheng commented:
@exekias I thought about the `mock` parameter as well as how should we pass in the aws credentials a little bit more. What I came up with is, using `aws_config_profile_name` to replace `mock` and aws credentials. Please let me know if this make sense :-)

`aws_config_profile_name` is the profile name in local AWS configuration file that the user want this aws metricbeat module to use. If `aws_config_profile_name` is not set, this module will report mock data instead. 

The AWS CLI stores the credentials that you specify with aws configure in a local file named credentials in a folder named .aws in your home directory `~/.aws/credentials`. The file generated by the AWS CLI for the profile configured in the previous section look like this:

> [default]
 aws_access_key_id=AKIAIOSFODNN7EXAMPLE
 aws_secret_access_key=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
[metricbeat-aws]
 aws_access_key_id=ASIAZENKQPPNZDBTSVNB
 aws_secret_access_key=Saxlp2wI09tMcaBhwkEMtWBjaGUwtrmGE
 aws_session_token=FQoGZXIvYXdzEHsaDKpiiMUFzVp16DT6MSKwAfm6efgVZWRcziaBt33Fn1Z3PvPfOOCXm0eWasKJlCenFLKfU2gaTLb0JursUeH+Af6qfJobz7MhpVYNbors7Ct59pOgYMRa8+3E1+3vW2iH7QDUdq/hIpK/kh03EotxMRb

Both `default` and `metricbeat-aws` in the example above are profile names. If we specify `aws_config_profile_name: "metricbeat-aws"`, then aws module will go to `~/.aws/credentials`and load config from profile `metricbeat-aws`.

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,313 @@
+				init = false
+				output, err := getMetricDataPerRegion(instanceID, getMetricDataOutput.NextToken, svcCloudwatch)
+				if err != nil {
+					report.Error(errors.Wrap(err, "getMetricDataPerRegion failed"))
```
you probably need to return here, maybe it could cause infinite loop if not?

kaiyan-sheng replied:
Nope this shouldn't be a `return` here. If 1 region failed, it should `continue` to the next region. Thanks for catching this!

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,313 @@
+	//actual fetch function
+	cfg, err := external.LoadDefaultAWSConfig(external.WithSharedConfigProfile(m.config.AwsConfigProfileName))
+	if err != nil {
+		logp.Err("Failed to load config: %s", err.Error())
```
`report.Error` + return?

exekias commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,313 @@
+	//Get a list of regions
+	regionsList, err := getRegions(svcEC2)
+	if err != nil {
+		report.Error(errors.Wrap(err, "getRegions failed"))
```
I would add the return here too, although this one should be safe, it removes uncertainty

exekias commented on x-pack/metricbeat/module/aws/ec2/data.go:
```diff
@@ -0,0 +1,41 @@
+)
+
+var (
+	s = schema.Schema{
```
I think you are not using this? Normally we use this to parse structured data (ie a JSON response) into desired field names

exekias commented:
That's an interesting approach, let's split the testing part from client settings.

Historically our users struggle with external files for configuration, as it requires more orchestration to configure. I think previous config was better as it allowed using config params, and if you want, pass these params through ENV vars. I've found a possible solution to avoid `SetEnv` and reading from files, I think this should work with the `aws` package:

```
awsAccessKey := "Insert Key Here" 
awsSecret := "Insert Secret Here" 
token := ""

creds := credentials.NewStaticCredentials(awsAccessKey, awsSecret, token)

cfg := aws.NewConfig().WithRegion("us-west-1").WithCredentials(creds)
```

Let me know what you think! I'll have a look to the mock and post again :)

kaiyan-sheng commented:
Thanks @exekias for your review!! @ph Just curious, did you use credentials in ~/.aws/credentials or environment variables in functionbeat?

kaiyan-sheng commented:
@exekias I did some more digging on `external.LoadDefaultAWSConfig` function in aws-go-sdk-v2. Turned out this function reads the SDK's default external configurations, and populates an AWS Config with the values from the external configurations and `default configuration sources are: * Environment Variables * Shared Configuration and Shared Credentials files`: https://godoc.org/github.com/aws/aws-sdk-go-v2/aws/external#LoadDefaultAWSConfig 
So with the ability to read from environment variables, I think we dont need `os.setenv` or `NewStaticCredentials` here as long as the environment variable `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY` and/or `AWS_SESSION_TOKEN` are set.

ruflin commented:
@kaiyan-sheng Can you check how much the addition of this module increases the size of our binary?

ruflin commented on x-pack/metricbeat/docs/modules/aws.asciidoc:
```diff
@@ -0,0 +1,65 @@
+////
```
As I can't comment on the image above, comment is here. The image above probably should go outside the x-pack directory into docs as all docs are in one place. As discussed before, we need to update docs_collector script to also collect the docs from the x-pack modules into the non-xpack docs directory: https://github.com/elastic/beats/blob/master/metricbeat/scripts/docs_collector.py

kaiyan-sheng replied:
I tried to move `x-pack/metricbeat/docs/modules/aws.asciidoc` into non-xpack directory `beats/metricbeat/docs/modules/aws.asciidoc` but seems like CI doesn't like it hmm https://travis-ci.org/elastic/beats/jobs/467148471

ruflin commented on x-pack/metricbeat/docs/modules_list.asciidoc:
```diff
@@ -0,0 +1,14 @@
+////
```
This should be added to the global list, see comment above.

ruflin commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,316 @@
+		return
+	}
+
+	cfg.Region = "us-west-1"
```
Will we have a default region configured?

kaiyan-sheng replied:
done! It's replaced by AWS_REGION as a part of the config. Thanks!

exekias commented:
This may be helpful so posting it here. I saw the API changes with sdk-v2, did some digging and probably this should work to pass user config as static credentials:

```
cfg := *awssdk.NewConfig()
cfg.Credentials = awssdk.StaticCredentialsProvider{
	Value: awssdk.Credentials{
		AccessKeyID:     m.config.AwsAccessKeyID,
		SecretAccessKey: m.config.AwsSecretAccessKey,
		SessionToken:    m.config.AwsSessionToken,
	},
}
cfg.Region = m.config.AwsDefaultRegion
```

exekias commented:
main reasons why I'm push for credentials in the beat config file are:
- They can be defined from central management
- This would enable users to store them in the beat [keystore](https://www.elastic.co/guide/en/beats/filebeat/current/keystore.html), to add more security

kaiyan-sheng commented:
@exekias Thanks for looking into using StaticCredentialsProvider. I tried this out several days ago and including today I'm still hitting a `invalid memory address or nil pointer dereference` error (my favorite golang error). If I replace https://github.com/elastic/beats/blob/1507cfa5f7f8e2f1ddb472a9b5bffcd2f2a83d3a/x-pack/metricbeat/module/aws/ec2/ec2.go#L109 with `cfg := *awssdk.NewConfig()`, I will get error:

> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1306ff4]
goroutine 5 [running]:
testing.tRunner.func1(0xc00019c300)
    /usr/local/go/src/testing/testing.go:792 +0x387
panic(0x1787100, 0x2204210)
    /usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/elastic/beats/vendor/github.com/aws/aws-sdk-go-v2/aws.New(0xc0000240ab, 0x9, 0x1998b60, 0xc0001e20c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /Users/kaiyansheng/go/src/github.com/elastic/beats/vendor/github.com/aws/aws-sdk-go-v2/aws/request.go:102 +0x124
github.com/elastic/beats/vendor/github.com/aws/aws-sdk-go-v2/aws.(*Client).NewRequest(0xc000059c00, 0xc00001f440, 0x178e820, 0xc00001f400, 0x17aff00, 0xc00000d340, 0x2212101)
    /Users/kaiyansheng/go/src/github.com/elastic/beats/vendor/github.com/aws/aws-sdk-go-v2/aws/client.go:73 +0x18a
github.com/elastic/beats/vendor/github.com/aws/aws-sdk-go-v2/service/ec2.(*EC2).newRequest(0xc00000e138, 0xc00001f440, 0x178e820, 0xc00001f400, 0x17aff00, 0xc00000d340, 0x40)
    /Users/kaiyansheng/go/src/github.com/elastic/beats/vendor/github.com/aws/aws-sdk-go-v2/service/ec2/service.go:72 +0x64
github.com/elastic/beats/vendor/github.com/aws/aws-sdk-go-v2/service/ec2.(*EC2).DescribeRegionsRequest(0xc00000e138, 0xc00001f400, 0x18c79e7, 0xc000059c00, 0x18b7daf)
    /Users/kaiyansheng/go/src/github.com/elastic/beats/vendor/github.com/aws/aws-sdk-go-v2/service/ec2/api.go:8666 +0x106
github.com/elastic/beats/x-pack/metricbeat/module/aws/ec2.getRegions(0x19acb40, 0xc00000e138, 0x1998b60, 0xc0001e20c0, 0x0, 0x0, 0x0)
    /Users/kaiyansheng/go/src/github.com/elastic/beats/x-pack/metricbeat/module/aws/ec2/ec2.go:180 +0x5c
github.com/elastic/beats/x-pack/metricbeat/module/aws/ec2.(*MetricSet).Fetch(0xc0001c85f0, 0x199a500, 0xc000191e00)
    /Users/kaiyansheng/go/src/github.com/elastic/beats/x-pack/metricbeat/module/aws/ec2/ec2.go:127 +0x29c
github.com/elastic/beats/metricbeat/mb/testing.ReportingFetchV2(0x5a391e8, 0xc0001c85f0, 0x176e9a0, 0xc000190d50, 0x5a391e8, 0xc0001c85f0, 0xf, 0xc000055538)
    /Users/kaiyansheng/go/src/github.com/elastic/beats/metricbeat/mb/testing/modules.go:203 +0x60
github.com/elastic/beats/x-pack/metricbeat/module/aws/ec2.TestFetch(0xc00019c300)
    /Users/kaiyansheng/go/src/github.com/elastic/beats/x-pack/metricbeat/module/aws/ec2/ec2_test.go:181 +0x554
testing.tRunner(0xc00019c300, 0x18daa40)
    /usr/local/go/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
    /usr/local/go/src/testing/testing.go:878 +0x353

`external.LoadDefaultAWSConfig` creates an aws.Config which is the same as `cfg := *awssdk.NewConfig()` or I tried even just `cfg := awssdk.Config{}`. Do you know maybe where I did wrong here?

ruflin commented on x-pack/metricbeat/module/aws/ec2/_meta/data.json:
```diff
@@ -0,0 +1,64 @@
+            }
+        }
+    },
+    "cloud": {
```
This should be mapped to ECS: https://github.com/elastic/ecs#cloud

kaiyan-sheng replied:
@ruflin I think I followed the ecs naming here already hmmm Sorry can you give me an example which part didn't map right please :-)

ruflin replied:
Sorry, my bad here. I thought it is under `aws` but it is on the top level. All good, sorry for the noise.

ruflin commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,325 @@
+	event := mb.Event{}
+	event.RootFields = common.MapStr{}
+	event.RootFields.Put("service.name", aws.ModuleName)
+	event.RootFields.Put("cloud.provider", "ec2")
```
Looks like the above generated event is just out of sync?

kaiyan-sheng replied:
@ruflin sorry what do you mean by out of sync here? This function `reportCloudWatchEvents` is to report one event per one specific instanceID.

ruflin replied:
All good, was related to my data.json comment. Ignore it :-)

kaiyan-sheng commented:
Regarding to binary size, @ruflin here is what I found:
Before adding aws module, metricbeat binary is about 54M.
> kaiyansheng@KaiyanMacBookPro:~/go/src/github.com/elastic/beats/metricbeat (aws_ec2)$ ls -lart | grep metricbeat
-rwxr-xr-x   1 kaiyansheng  staff    54M Dec 13 13:46 metricbeat*
With aws module, metribeat binary size went up significantly:
kaiyansheng@KaiyanMacBookPro:~/go/src/github.com/elastic/beats/metricbeat (aws_ec2)$ ls -lart | grep metricbeat
-rwxr-xr-x   1 kaiyansheng  staff    68M Dec 13 18:46 metricbeat*

ruflin commented:
@kaiyan-sheng I'm really surprised by the massive increase in size as I think it cannot be explained by just the 60k lines added. Could it be that it pulls in stuff from the core lib which we don't use yet? Any chance you could take a look on what makes this massive difference?

One thing that surprises me in your code above: It seems you are buildng the OSS binary and no the x-pack one? What is the exact command you used for building the binary?

exekias commented:
> @exekias Thanks for looking into using StaticCredentialsProvider. I tried this out several days ago and including today I'm still hitting a `invalid memory address or nil pointer dereference` error (my favorite golang error). If I replace
> 
> [beats/x-pack/metricbeat/module/aws/ec2/ec2.go](https://github.com/elastic/beats/blob/1507cfa5f7f8e2f1ddb472a9b5bffcd2f2a83d3a/x-pack/metricbeat/module/aws/ec2/ec2.go#L109)

You are right! it seems some more initialization is needed :innocent:, I found this: https://godoc.org/github.com/aws/aws-sdk-go-v2/aws/defaults

so `cfg := defaults.Config()` may do the trick, it seems to work on my side

elasticmachine commented:
Pinging @elastic/infrastructure

ruflin commented:
@kaiyan-sheng If you add the imports manually to the list in x-pack I would expect that you can just run `make` or `go build` there? Do you see then the same diff?

ruflin commented on vendor/github.com/aws/aws-sdk-go-v2/service/cloudwatch/service.go:
```diff
@@ -0,0 +1,80 @@
+
+import (
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/signer/v4"
```
@kaiyan-sheng I had a look at the imports and compared them to what is checked in here. It seems for example this package here is not checked in? Could it be that you have it locally but not versioned? This could explain why the binary is so big compared to the changed code lines.

exekias replied:
I don't think the compiler will put anything in the binary if it's not being used. For example, when you remove the module from `include.go` the resulting binary goes back to ~65MB.

Do you know how we can debug this to understand the reason for this extra ~15MB? Perhaps @urso knows this?

@ruflin how concerned are you about this? Perhaps we can go for it's own beat if you think it's a big deal

kaiyan-sheng replied:
@ruflin hmm for aws/signer/v4? This is already in https://github.com/elastic/beats/blob/master/vendor/vendor.json#L231 and https://github.com/elastic/beats/tree/master/vendor/github.com/aws/aws-sdk-go-v2/aws/signer/v4

ruflin replied:
@kaiyan-sheng Got it, that is why some of the dependency are missing.

@exekias As I assume adding future metricsets to it will not increase the size I'm ok with getting it in. But I would still like to understand why it increases much more then I think it should.

ruflin commented on x-pack/metricbeat/module/aws/ec2/ec2_integration_test.go:
```diff
@@ -0,0 +1,119 @@
+		for _, event := range events {
+			fmt.Println("event = ", event)
+			// RootField
+			checkEventField("service.name", "string", event, t)
```
Alternatively what I think should also work here is creating a `common.MapStr` with all the fields inside and then compare it. Something like:

```
result := common.MapStry{
  "cloud": common.MapStr{
    "instance": common.MapStr{
      "id": "foo"
    }
  }
}
```

The type would then also be directly compared if it is correct.

kaiyan-sheng replied:
@ruflin Your suggestion will work for TestMockFetch in ec2_test.go but probably won't work here in integration test though. In this integration test, it will go to an actual aws account to query for ec2 instance information and cloudwatch metrics with real data which the value will vary. Am I misunderstanding your suggestion?

ruflin replied:
No, that sounds correct.

kaiyan-sheng commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,393 @@
+		Value: &instanceID,
+	}
+
+	metricDataQuery1 := createMetricDataQuery("cpu1", "CPUUtilization", periodInSec, []cloudwatch.Dimension{dim})
```
Great idea! That will make this code a lot cleaner!  Thanks!

kaiyan-sheng commented on x-pack/metricbeat/module/aws/_meta/fields.yml:
```diff
@@ -0,0 +1,12 @@
+  title: "aws"
+  description: >
+    `aws` module collects AWS monitoring metrics from AWS Cloudwatch.
+    experimental[]
```
Will do!

kaiyan-sheng commented on x-pack/metricbeat/module/aws/_meta/docs.asciidoc:
```diff
@@ -0,0 +1,36 @@
+a secret access key, and a security token which typically returned from `GetSessionToken`. MFA-enabled IAM users would
+need to submit an MFA code while calling `GetSessionToken`. `aws_default_region` is to set the region for SDK to use.
+
+If `aws_access_key_id` or `aws_secret_access_key` not set, this module will report mock data instead.
```
Sorry this should be removed. Not applicable any more.

kaiyan-sheng commented on x-pack/metricbeat/module/aws/_meta/docs.asciidoc:
```diff
@@ -0,0 +1,36 @@
+`access keys` are long-term credentials for an IAM user or the AWS account root user. A more AWS recommended way is to use
+`temporary security credentials` instead of `access keys`. `temporary security credentials` consist of an access key ID,
+a secret access key, and a security token which typically returned from `GetSessionToken`. MFA-enabled IAM users would
+need to submit an MFA code while calling `GetSessionToken`. `aws_default_region` is to set the region for SDK to use.
```
@ruflin If the user is using `temporary security credentials` then yes, once the token expired, a new token needs to be inserted. Otherwise aws module will stop reporting metrics. Is this approach ok?

ruflin replied:
I don't think we really have an alternative here. What I worry is that the user has to restart Metricbeat in this case except he uses reloading of the module.

What is our recommended lifetime for the session key?

Do we log a good error when the key expires?

kaiyan-sheng commented on x-pack/metricbeat/module/aws/_meta/docs.asciidoc:
```diff
@@ -0,0 +1,36 @@
+This module environment variable `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_SESSION_TOKEN` and `AWS_REGION`
+references in the config file to set values that need to be configurable during deployment.
+
+There are two different kinds of AWS credentials can be used here: `access keys` and `temporary security credentials`.
```
I added some links and aws sts command. Thanks!

kaiyan-sheng commented on x-pack/metricbeat/module/aws/aws.go:
```diff
@@ -0,0 +1,48 @@
+// Config defines all required and optional parameters for aws metricsets
+type Config struct {
+	Period             string `config:"period"`
+	AwsAccessKeyID     string `config:"aws_access_key_id"`
```
good point, I will remove `aws` prefix since it's a bit redundant.

kaiyan-sheng commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,393 @@
+		return
+	}
+
+	for _, regionName := range regionsList {
```
This is not the configured `aws_default_region`, this regionsList is a list of all the region names returned from `getRegions` function.

ruflin replied:
I think I miss something here. Why do we iterated through all regions if the users has only configured 1 region?

kaiyan-sheng replied:
The one configured region that user put in is only a part of the credentials to use to make aws api  calls. That is not the regions that user wants to monitor.

kaiyan-sheng commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,393 @@
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+	//mock Fetch
```
Let's talk about this before I make any more changes :-) With the current code, if user forgot to configure aws credentials/keys, then error `2019-01-10T15:54:02.289-0700    ERROR   instance/beat.go:907    Exiting: 1 error: missing field accessing '0.access_key_id' (source:'/Users/kaiyansheng/go/src/github.com/elastic/beats/x-pack/metricbeat/modules.d/aws.yml')
Exiting: 1 error: missing field accessing '0.access_key_id' (source:'/Users/kaiyansheng/go/src/github.com/elastic/beats/x-pack/metricbeat/modules.d/aws.yml')
` will show up in the log and metricbeat will stop. I don't think this is the right behavior either.

kaiyan-sheng commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,393 @@
+
+	svcEC2 := ec2.New(cfg)
+	//Get a list of regions
+	regionsList, err := getRegions(svcEC2)
```
Good point. I don't think we need to fetch all the region names every 5 minutes since they will not change very often.

ruflin replied:
How often should we refresh them? Once a day? Or even cache them forever by default and have a config option to turn this off?

kaiyan-sheng replied:
@ruflin Right now it caches the region list forever till next time metricbeat gets restarted. I think in the future, maybe making regionList configurable can be considered. In that case, users can specify what are the regions they are interested to monitor. What you think about this?

ruflin commented on x-pack/metricbeat/module/aws/ec2/_meta/data.json:
```diff
@@ -0,0 +1,67 @@
+    },
+    "service": {
+        "name": "aws",
+        "type": "aws"
```
How did you try to set it? There is a variable `Service in the `mb.Event` struct.

kaiyan-sheng replied:
@ruflin ahhh thanks! I figured out after reading the mb.Event code! Thanks!

ruflin commented on x-pack/metricbeat/module/aws/_meta/docs.asciidoc:
```diff
@@ -0,0 +1,54 @@
+----
+
+By default, Amazon EC2 sends metric data to CloudWatch every 5 minutes. With this basic monitoring, `period` in aws module
+configuration should be larger or equal than `300s`. If `period` is set to be less than `300s`, the same cloudwatch metrics
```
I assume in the default on 300s or a multiple of 300s make sense? So if I set it to 400s I get sometimes the same data twice and some data points I could miss?

kaiyan-sheng replied:
@ruflin If set it to 400s, some data points will be missed. I added some recommendation in this doc. Thanks!

kaiyan-sheng commented:
@ruflin @exekias After the conversation we had earlier today, I moved `getRegions` into `New` function so it will only be called once when starting, instead of getting called every collection period. 

Also, when I was working on removing `MockFetch` function, I realized `TestMockFetch` is actually just testing `createCloudWatchEvents` function. So instead of making everything complicated, I removed `MockFetch` function and replaced `TestMockFetch` in unit test with `TestCreateCloudWatchEvents`. Please let me know if this is OK for you or if you have any concerns. Thanks for all the advice!

ruflin commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -79,61 +71,61 @@ var metricIDNameMap = map[string][]string{
+	for _, regionName := range m.regionsList {
+		fmt.Println("regionName = ", regionName)
+		m.awsConfig.Region = regionName
+		// cfg.Region = regionName
```
leftover?

kaiyan-sheng replied:
@ruflin yes done! Thanks!

ruflin commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -79,61 +71,61 @@ var metricIDNameMap = map[string][]string{
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+	for _, regionName := range m.regionsList {
+		fmt.Println("regionName = ", regionName)
```
Leftover?

kaiyan-sheng replied:
@ruflin yes! Thanks! Done!

exekias commented:
LGTM :+1:

kaiyan-sheng commented on x-pack/metricbeat/module/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,393 @@
+	)
+}
+
+// MockEC2Client struct is used for unit tests.
```
done

ruflin commented:
I did some more testing on what happens if an EC2 instance is added and one deleted. It detects the new instance and stops collecting data for the old one. The problem with the new instance is that the first events don't have stats inside until we have it in CloudWatch. (yes, I used a low period).

ruflin commented:
@kaiyan-sheng Please rebase on top of https://github.com/elastic/beats/pull/10100 to get the build green.

Would also be nice to update the PR description a bit and format it. It should contain a bit more details on what ended up in this PR. Can you also create a meta issue for the complete AWS module with links to the other metricsets issues so we have a central place to track it?

kaiyan-sheng commented:
Sounds good. Thanks for testing it and for all the comments. I created a meta issue for this and also added a checklist for this PR. I will merge this PR and create a new PR to address the rest of the comments.

kaiyan-sheng commented on x-pack/metricbeat/module/aws/_meta/docs.asciidoc:
```diff
@@ -0,0 +1,56 @@
+`sts get-session-token` AWS CLI can be used to generate temporary credentials. For example. with MFA-enabled:
+
+----
+aws> sts get-session-token --serial-number arn:aws:iam::1234:mfa/test@elastic.co --token-code 456789 --duration-seconds 129600
```
@ruflin Unfortunately 36 hour is the maximum for temporary token. `The GetSessionToken action must be called by using the long-term AWS security credentials of the AWS account or an IAM user. Credentials that are created by IAM users are valid for the duration that you specify, from 900 seconds (15 minutes) up to a maximum of 129600 seconds (36 hours), with a default of 43200 seconds (12 hours).`

kaiyan-sheng replied:
https://docs.aws.amazon.com/cli/latest/reference/sts/get-session-token.html

ruflin replied:
I see. Should we even recommend to use these temporary tokens?
